app/models/application_record.rb

```diff
@@ -13,4 +13,12 @@ def self.logger
   def logger
     self.class.logger
   end
+
+  def self.graphql_type(new_graphql_type = nil)
+    if new_graphql_type
+      @graphql_type = new_graphql_type
+    else
+      @graphql_type || (superclass.respond_to?(:graphql_type) ? superclass.public_send(:graphql_type) : nil)
```

Comment:
This would recurse up the class tree, either until getting to ApplicationRecord, which will return `nil` because ActiveRecord doesn't respond to `graphql_type`, or return the class instance variable along the way if it was set. Which would happen anyways if the function isn't defined on the child class.
So perhaps we could simplify the implementation greatly by having
```ruby
def self.graphql_type
  nil
end
```
here, and override it only on the classes you want to with something similar to:
```ruby
def self.graphql_type
  ::Types::Host
end
```
and that way we don't need to work with passing lambdas around and calling them in the schema etc

Reply:
> This would recurse up the class tree, either until getting to ApplicationRecord, which will return `nil` because ActiveRecord doesn't respond to `graphql_type`, or return the class instance variable along the way if it was set.

Yes, that was the aim of this. I don't see a problem this to be honest.

> override it only on the classes you want to with something similar to:

I really like having a DSL for this. That way it's very active record style. Overriding methods does work, but does not imply `this is a stable API developers can use and rely on`.

> work with passing lambdas around

We just need the lambdas to solve a circular dependency error. The types reference the classes and the classes the types. That did not work in my testing.

Reply:
Could we just use strings instead of lambdas to prevent the circular dependency? then `safe_constantize` would work for that case and reduce the complexity. I'm trying to avoid lambdas if possible because they make the code more difficult to understand.
i think you can also simplify the retrieval with:
```suggestion
      @graphql_type || superclass.try(:graphql_type)
```

Reply:
> Could we just use strings instead of lambdas to prevent the circular dependency?

Yes, we definitely can. I actually implemented this in the first iteration but decided to go for lambdas as I think Strings are pretty misleading. I know rails usually supports both strings and lambdas. How about we do the following: I change this to strings and open a follow-up PR that adds supports for lambdas (with proper test coverage etc.). WDYT?

Reply:
I also want to avoid `try` if possible as this swallows all exceptions which can make debugging harder. I know the code is harder to read but if it helps find a bug where is happens, it's worth it imho.

Reply:
@tbrisker: What do you think about adding an abstraction layer like this:

```ruby
module Foreman
  module Module
    # This method accepts either a string or a lambda and resolves it to a constant
    def self.resolve(mod)
      mod = mod.call if mod.respond_to?(:call)
      return mod.constantize if mod.instance_of?(String)
      mod
    end
  end
end
```

We can both support strings and lambdas and add proper testing for this.